lib/src/git.rs

```diff
@@ -1319,12 +1326,42 @@ fn push_refs(
             GitPushError::InternalGitError(err)
         }
     })?;
-    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs.iter().copied().collect();
+    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs_expected_locations
+        .keys()
+        .copied()
+        .collect();
     let mut push_options = git2::PushOptions::new();
     let mut proxy_options = git2::ProxyOptions::new();
     proxy_options.auto();
     push_options.proxy_options(proxy_options);
     let mut callbacks = callbacks.into_git();
+    callbacks.push_negotiation(|updates| {
+        for update in updates {
+            let dst_refname = update
+                .dst_refname()
+                .expect("Expect reference name to be valid UTF-8");
+            let expected_remote_location = *qualified_remote_refs_expected_locations
+                .get(dst_refname)
+                .expect("Push is trying to move a ref it wasn't asked to move");
+            let expected_location_oid = expected_remote_location
+                .map(|id| git2::Oid::from_bytes(id.as_bytes()).unwrap())
+                .unwrap_or(git2::Oid::zero());
```

Comment:
nit: conversion to `git2::Oid` isn't needed anymore (and rewritten by the subsequent commit.)